
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Welcome to PrimerSeq! &mdash; PrimerSeq 1.0.0 documentation</title>
    
    <link rel="stylesheet" href="_static/flasky.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '',
        VERSION:     '1.0.0',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="top" title="PrimerSeq 1.0.0 documentation" href="#" />
   
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9">

  </head>
  <body>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li><a href="#">PrimerSeq 1.0.0 documentation</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <div class="section" id="welcome-to-primerseq">
<h1>Welcome to PrimerSeq!<a class="headerlink" href="#welcome-to-primerseq" title="Permalink to this headline">¶</a></h1>
<p>PrimerSeq aims to design RT-PCR primers on appropriate flanking
exons around your target of interest. Its aim is to design
primers for validating alternative splicing events. PrimerSeq is particularly
advantageous for designing a large number of primers for validating differential
alternative splicing events found in RNA-Seq data.</p>
<p>PrimerSeq is licensed under <a class="reference external" href="licenses/gpl-3.0.txt">GPLv3</a>.
PrimerSeq relies on the following software:</p>
<ul class="simple">
<li>Primer3 - <a class="reference external" href="licenses/gpl-2.0.txt">GPLv2</a></li>
<li>SAM-JDK - <a class="reference external" href="licenses/apache-license-2.0.txt">Apache License V2.0</a>, <a class="reference external" href="licenses/mit-license.txt">MIT</a></li>
<li>BigWig Java API - <a class="reference external" href="licenses/lgpl-2.1.txt">LGPL v2.1</a></li>
</ul>
</div>
<div class="section" id="external-documentation">
<h1>External Documentation<a class="headerlink" href="#external-documentation" title="Permalink to this headline">¶</a></h1>
<ul class="simple">
<li><a class="reference external" href="http://primerseq.sf.net/getting_started.html">Getting Started</a></li>
<li><a class="reference external" href="http://primerseq.sf.net/user_guide.html">User Tutorial</a></li>
<li><a class="reference external" href="http://primerseq.sf.net/gtf.html">GTF Documentation</a></li>
<li><a class="reference external" href="http://primerseq.sf.net/faq.html">FAQ</a></li>
</ul>
</div>
<div class="section" id="user-documentation">
<h1>User Documentation<a class="headerlink" href="#user-documentation" title="Permalink to this headline">¶</a></h1>
<p>The PrimerSeq Graphical User Interface (GUI) consists of four main components.</p>
<ol class="arabic simple">
<li>Menu items</li>
<li>Required tab</li>
<li>Optional tab</li>
<li>Results Window</li>
</ol>
<div class="section" id="menu-items">
<h2>1. Menu items<a class="headerlink" href="#menu-items" title="Permalink to this headline">¶</a></h2>
<p>The menu has several import options within three categories.</p>
<ol class="arabic simple">
<li>File</li>
<li>Edit</li>
<li>Help</li>
</ol>
<div class="section" id="file">
<h3>1.1 File<a class="headerlink" href="#file" title="Permalink to this headline">¶</a></h3>
<div class="section" id="load-ex">
<h4>Load Ex.<a class="headerlink" href="#load-ex" title="Permalink to this headline">¶</a></h4>
<p>The File-&gt;Load Ex. option automatically loads example data that is distributed with PrimerSeq.
For details read the <a class="reference external" href="http://primerseq.sf.net/getting_started.html">getting started</a> page.</p>
</div>
<div class="section" id="reset">
<h4>Reset<a class="headerlink" href="#reset" title="Permalink to this headline">¶</a></h4>
<p>The File-&gt;Reset option resets PrimerSeq in case an error occurs during running.</p>
</div>
<div class="section" id="quit">
<h4>Quit<a class="headerlink" href="#quit" title="Permalink to this headline">¶</a></h4>
<p>The File-&gt;Quit option closes PrimerSeq.</p>
</div>
</div>
<div class="section" id="edit">
<h3>2.1 Edit<a class="headerlink" href="#edit" title="Permalink to this headline">¶</a></h3>
<div class="section" id="primer3">
<h4>Primer3<a class="headerlink" href="#primer3" title="Permalink to this headline">¶</a></h4>
<p>The Edit-&gt;Primer3 option opens a configuration file for Primer3 called primer3.txt. For information on
editing Primer3 parameters please see their <a class="reference external" href="../primer3/primer3_manual.htm">manual</a>.</p>
</div>
<div class="section" id="sort-gtf-add-genes">
<h4>Sort GTF &amp; Add Genes<a class="headerlink" href="#sort-gtf-add-genes" title="Permalink to this headline">¶</a></h4>
<p>For more information about these options, please refer to the <a class="reference external" href="http://primerseq.sf.net/gtf.html">GTF documentation</a>.</p>
</div>
</div>
</div>
<div class="section" id="required-tab">
<h2>2. Required tab<a class="headerlink" href="#required-tab" title="Permalink to this headline">¶</a></h2>
<p>The required tab consists of input that must be filled in before PrimerSeq can design
primers. The required input is as follows:</p>
<ol class="arabic simple">
<li>Genome sequence (FASTA)</li>
<li>Gene annotation (GTF)</li>
<li>Mapped reads (SAM/BAM(s))</li>
<li>Target exon(s)</li>
<li>Output file</li>
</ol>
<img alt="alternate text" class="align-center" src="_images/required.png" />
<div class="section" id="genome-sequence-fasta">
<h3>2.1 Genome sequence (FASTA)<a class="headerlink" href="#genome-sequence-fasta" title="Permalink to this headline">¶</a></h3>
<p>Genome sequences are only allowed in FASTA format. Use FASTA files that contain the entire genome
sequence for a species to avoid problems. You can obtain the Human and Mouse genome from the following
links:</p>
<ul class="simple">
<li><a class="reference external" href="http://hgdownload.cse.ucsc.edu/goldenPath/hg19/bigZips/hg19.2bit">hg19</a></li>
<li><a class="reference external" href="http://hgdownload.cse.ucsc.edu/goldenPath/mm9/bigZips/mm9.2bit">mm9</a></li>
</ul>
<p>The first time you load a FASTA file may take several minutes. Subsequent loading will occur instantly.</p>
</div>
<div class="section" id="gene-annotation-gtf">
<h3>2.2 Gene annotation (GTF)<a class="headerlink" href="#gene-annotation-gtf" title="Permalink to this headline">¶</a></h3>
<p>PrimerSeq uses gene annotation in GTF format. Ultimately, the choice of where you obtain the GTF is up to you. You could download GTFs known to work from the PrimerSeq sourceforge website, download GTFs from UCSC or Ensembl, or use GTF output from transcript assemblers like Cufflinks. For ease of use, GTFs for human and mouse are provided. UCSC&#8217;s known gene annotation is preferred over Ensembl.</p>
<ul class="simple">
<li><a class="reference external" href="http://sourceforge.net/projects/primerseq/files/GTF/Homo_sapiens.knownGene.hg19.sorted.withGenes.gtf.gz/download">UCSC Human (hg19)</a></li>
<li><a class="reference external" href="http://sourceforge.net/projects/primerseq/files/GTF/Homo_sapiens.GRCh37.69.sorted.gtf.gz/download">Ensembl Human (hg19)</a></li>
<li><a class="reference external" href="http://sourceforge.net/projects/primerseq/files/GTF/Mus_musculus.knownGene.mm9.sorted.withGenes.gtf.gz/download">UCSC Mouse (mm9)</a></li>
<li><a class="reference external" href="http://sourceforge.net/projects/primerseq/files/GTF/Mus_musculus.GRCm38.69.sorted.gtf.gz/download">Ensembl Mouse (mm9)</a></li>
</ul>
<p>You can also mix several GTF files into a single input GTF file for PrimerSeq! For more detailed explanations please click <a class="reference external" href="http://primerseq.sourceforge.net/gtf.html">here</a>.</p>
</div>
<div class="section" id="mapped-reads-sam-bam-s">
<h3>2.3 Mapped reads (SAM/BAM(s))<a class="headerlink" href="#mapped-reads-sam-bam-s" title="Permalink to this headline">¶</a></h3>
<p>You can specify none, one or multiple SAM/BAM files as input. Make sure to hold down the ctrl key to select multiple files. SAM/BAM files are the typical output format from read aligners like Tophat. The SAM/BAM file allows PrimerSeq to estimate the relative abundance of different isoforms. If you know your BAM file is sorted (like the output from Tophat) then please name the BAM file with a .sorted.bam extension so PrimerSeq does not try to re-sort the file.</p>
</div>
<div class="section" id="target-exon-s">
<h3>2.4 Target exon(s)<a class="headerlink" href="#target-exon-s" title="Permalink to this headline">¶</a></h3>
<p>Enter target coordinates as (strand)chr:start-end. The first nucleotide on a chromosome is at position 0. The end position is not inclusive.</p>
<p>If you wish PrimerSeq to automatically find flanking exons then only specify a list of target exons where each exon is specified on a different line.</p>
<p>-chr18:9562919-9563044</p>
<p>However, you may also specify the flanking exons for primer design. Follow the following convention to do this:</p>
<p>targetExon1,upstreamExon1,downstreamExon1
targetExon2,upstreamExon2,downstreamExon2</p>
<p>The target exon is followed by upstream flanking exon and then by the downstream flanking exon (separated by a comma).
Remember to follow the same coordinate format for just the target exon case.</p>
</div>
<div class="section" id="output-file">
<h3>2.5 Output file<a class="headerlink" href="#output-file" title="Permalink to this headline">¶</a></h3>
<p>PrimerSeq saves results to the text file you specify immediately after primer design is finished.</p>
</div>
</div>
<div class="section" id="optional-tab">
<h2>3. Optional tab<a class="headerlink" href="#optional-tab" title="Permalink to this headline">¶</a></h2>
<img alt="alternate text" class="align-center" src="_images/optional.png" />
<ul class="simple">
<li><strong>Minimum Flanking PSI:</strong> minimum inclusion level where a primer is allowed to be placed on that exon</li>
<li><strong>Splice Junction:</strong> source for defining exon junctions</li>
<li><strong>Gene ID:</strong> flag for whether the GTF has a valid Gene ID</li>
<li><strong>Keep Temporary:</strong> flag for keeping temporary files created when running PrimerSeq</li>
<li><strong>Read Threshold:</strong> minimum read count necessary to define a novel junction from RNA-Seq</li>
<li><strong>Anchor Length:</strong> minimum number of bases on both sides of a junction for a read to be considered valid</li>
<li><strong>Min. Jct Count:</strong> assign at least this number of reads to a junction known in the GTF annotation</li>
</ul>
</div>
<div class="section" id="results-window">
<h2>4. Results Window<a class="headerlink" href="#results-window" title="Permalink to this headline">¶</a></h2>
<img alt="alternate text" class="align-center" src="_images/display_results.png" />
<p>The results window allows you to plot the results of primer design and check for unintended product
amplification using UCSC&#8217;s In-Silico PCR. For information on producing these results, please see the
<a class="reference external" href="http://primerseq.sf.net/getting_started.html">getting started</a> page.</p>
<img alt="alternate text" class="align-center" src="_images/display_plot.png" />
<img alt="alternate text" class="align-center" src="_images/insilico_pcr_results.png" />
<div class="toctree-wrapper compound">
<ul class="simple">
</ul>
</div>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper"><h1>PrimerSeq</h1>
<p>PrimerSeq aims to design RT-PCR primers which
validate alternative splicing events.</p>

<h1>Feedback</h1>
<p>If you have suggestions please contact 
<a href="mailto:primerseq@gmail.com">Collin Tokheim</a>.
You can also raise issues on <a href="https://github.com/ctokheim/PrimerSeq/issues/">GitHub</a></p>
  <h3>This Page</h3>
  <ul class="this-page-menu">
    <li><a href="_sources/index.txt"
           rel="nofollow">Show Source</a></li>
  </ul>
<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy; Copyright 2012, CollinTokheim.
    </div>
    <a href="https://github.com/ctokheim/PrimerSeq" class="github">
        <img style="position: absolute; top: 0; right: 0; border: 0;" src="http://s3.amazonaws.com/github/ribbons/forkme_right_darkblue_121621.png" alt="Fork me on GitHub" />
    </a>
<!--
    <script type="text/javascript">

      var _gaq2 = _gaq2 || [];
      _gaq2.push(['_setAccount', 'UA-8742933-10']);
      _gaq2.push(['_setDomainName', 'none']);
      _gaq2.push(['_setAllowLinker', true]);
      _gaq2.push(['_trackPageview']);

      (function() {
        var ga2 = document.createElement('script'); ga.type = 'text/javascript'; ga2.async = true;
        ga2.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
        var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga2, s);
      })();

    </script>

    <script type="text/javascript">
      (function() {
        var t   = document.createElement('script');
        t.type  = 'text/javascript';
        t.async = true;
        t.id    = 'gauges-tracker';
        t.setAttribute('data-site-id',
                       '4ddc1cfaf5a1f50fcc000001');
        t.src = '//secure.gaug.es/track.js';
        var s = document.getElementsByTagName('script')[0];
        s.parentNode.insertBefore(t, s);
      })();
    </script>
-->
  </body>
</html>