
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>splice_graph.py &mdash; PrimerSeq 1.0.0 documentation</title>
    
    <link rel="stylesheet" href="_static/flasky.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '',
        VERSION:     '1.0.0',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="author" title="About these documents" href="about.html" />
    <link rel="top" title="PrimerSeq 1.0.0 documentation" href="index.html" />
    <link rel="up" title="Code" href="code.html" />
    <link rel="next" title="algorithms.py" href="algorithms.html" />
    <link rel="prev" title="primer.py" href="primer.html" />
   
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9">

  </head>
  <body>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="algorithms.html" title="algorithms.py"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="primer.html" title="primer.py"
             accesskey="P">previous</a> |</li>
        <li><a href="index.html">PrimerSeq 1.0.0 documentation</a> &raquo;</li>
          <li><a href="code.html" accesskey="U">Code</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <div class="section" id="module-splice_graph">
<span id="splice-graph-py"></span><h1>splice_graph.py<a class="headerlink" href="#module-splice_graph" title="Permalink to this headline">¶</a></h1>
<p><strong>Author:</strong> Collin Tokheim</p>
<div class="section" id="splice-graph">
<h2>Splice Graph<a class="headerlink" href="#splice-graph" title="Permalink to this headline">¶</a></h2>
<p>The splice_graph.py deals with gene structure as a (weighted) directed
acyclic graph (DAG) normally known as a splice graph. Like you would expect,
the <a class="reference internal" href="#splice_graph.SpliceGraph" title="splice_graph.SpliceGraph"><tt class="xref py py-class docutils literal"><span class="pre">SpliceGraph</span></tt></a> class represents gene structure
as a splice graph.</p>
</div>
<div class="section" id="transcripts-overlapping-target">
<h2>Transcripts Overlapping Target<a class="headerlink" href="#transcripts-overlapping-target" title="Permalink to this headline">¶</a></h2>
<p><a class="reference internal" href="#splice_graph.get_from_gtf_using_gene_name" title="splice_graph.get_from_gtf_using_gene_name"><tt class="xref py py-func docutils literal"><span class="pre">get_from_gtf_using_gene_name()</span></tt></a>
returns the all transcripts from a gene if the gene overlaps the user&#8217;s
target. <a class="reference internal" href="#splice_graph.get_weakly_connected_tx" title="splice_graph.get_weakly_connected_tx"><tt class="xref py py-func docutils literal"><span class="pre">get_weakly_connected_tx()</span></tt></a> returns all transcripts
that are weakly connected to the user&#8217;s target. This approach is used when
gene IDs are not valid and thus can not be used.</p>
</div>
<div class="section" id="flanking-exons">
<h2>Flanking Exons<a class="headerlink" href="#flanking-exons" title="Permalink to this headline">¶</a></h2>
<p>If read counts are used (psi &lt; 1) then splice_graph uses the
<a class="reference internal" href="#splice_graph.get_sufficient_psi_exons" title="splice_graph.get_sufficient_psi_exons"><tt class="xref py py-func docutils literal"><span class="pre">get_sufficient_psi_exons()</span></tt></a> function to find
flanking exons with at least a user defined inclusion level.
If psi == 1 then flanking exons are only determined by the biconnected
componenets algorithm using <a class="reference internal" href="#splice_graph.get_flanking_biconnected_exons" title="splice_graph.get_flanking_biconnected_exons"><tt class="xref py py-func docutils literal"><span class="pre">get_flanking_biconnected_exons()</span></tt></a>.</p>
</div>
<div class="section" id="documentation">
<h2>Documentation<a class="headerlink" href="#documentation" title="Permalink to this headline">¶</a></h2>
<dl class="class">
<dt id="splice_graph.SpliceGraph">
<em class="property">class </em><tt class="descclassname">splice_graph.</tt><tt class="descname">SpliceGraph</tt><big>(</big><em>annotation</em>, <em>chr</em>, <em>strand</em>, <em>read_threshold=5</em>, <em>filter_factor=2</em>, <em>min_count=1</em><big>)</big><a class="reference internal" href="_modules/splice_graph.html#SpliceGraph"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#splice_graph.SpliceGraph" title="Permalink to this definition">¶</a></dt>
<dd><p>The SpliceGraph class is meant to provide a common repository for creation
of splice graphs for a single gene.</p>
<dl class="method">
<dt id="splice_graph.SpliceGraph.add_all_possible_edge_weights">
<tt class="descname">add_all_possible_edge_weights</tt><big>(</big><em>weights</em><big>)</big><a class="reference internal" href="_modules/splice_graph.html#SpliceGraph.add_all_possible_edge_weights"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#splice_graph.SpliceGraph.add_all_possible_edge_weights" title="Permalink to this definition">¶</a></dt>
<dd><p>Add edge/weights to graph if supported by atleast READ_THRESHOLD number of reads</p>
</dd></dl>

<dl class="method">
<dt id="splice_graph.SpliceGraph.set_annotation_edge_weights">
<tt class="descname">set_annotation_edge_weights</tt><big>(</big><em>weights</em><big>)</big><a class="reference internal" href="_modules/splice_graph.html#SpliceGraph.set_annotation_edge_weights"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#splice_graph.SpliceGraph.set_annotation_edge_weights" title="Permalink to this definition">¶</a></dt>
<dd><p>Only try to find weights for already existing edges in the graph.
This function is intended to add weight values to edges defined
in the gtf annotation.</p>
</dd></dl>

<dl class="method">
<dt id="splice_graph.SpliceGraph.set_graph_as_annotation">
<tt class="descname">set_graph_as_annotation</tt><big>(</big><em>annotation</em><big>)</big><a class="reference internal" href="_modules/splice_graph.html#SpliceGraph.set_graph_as_annotation"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#splice_graph.SpliceGraph.set_graph_as_annotation" title="Permalink to this definition">¶</a></dt>
<dd><p>Create a nx DiGraph from list of tx in gene. FILTER_FACTOR defines a cutoff for
using a tx of a gene. A tx must have x num of exon where x &gt; MAX tx exon num / FILTER_FACTOR.</p>
</dd></dl>

<dl class="method">
<dt id="splice_graph.SpliceGraph.set_graph_as_nodes_only">
<tt class="descname">set_graph_as_nodes_only</tt><big>(</big><em>exons</em><big>)</big><a class="reference internal" href="_modules/splice_graph.html#SpliceGraph.set_graph_as_nodes_only"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#splice_graph.SpliceGraph.set_graph_as_nodes_only" title="Permalink to this definition">¶</a></dt>
<dd><p>Simple function that makes a DAG (nx.DiGraph) with only nodes and no edges.
Meant to be used to before add_all_possible_edge_weights.</p>
</dd></dl>

</dd></dl>

<dl class="function">
<dt id="splice_graph.calculate_target_psi">
<tt class="descclassname">splice_graph.</tt><tt class="descname">calculate_target_psi</tt><big>(</big><em>target</em>, <em>sg_list</em>, <em>component</em><big>)</big><a class="headerlink" href="#splice_graph.calculate_target_psi" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculate psi for the target exon for each bam file.</p>
</dd></dl>

<dl class="function">
<dt id="splice_graph.get_flanking_biconnected_exons">
<tt class="descclassname">splice_graph.</tt><tt class="descname">get_flanking_biconnected_exons</tt><big>(</big><em>name</em>, <em>target</em>, <em>sGraph</em>, <em>genome</em><big>)</big><a class="reference internal" href="_modules/splice_graph.html#get_flanking_biconnected_exons"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#splice_graph.get_flanking_biconnected_exons" title="Permalink to this definition">¶</a></dt>
<dd><p>Defines flanking exons as exons that cannot be skipped in
the graph structure. Theese exons are 100% included and do not
need estimation of inclusion level.</p>
</dd></dl>

<dl class="function">
<dt id="splice_graph.get_from_gtf_using_gene_name">
<tt class="descclassname">splice_graph.</tt><tt class="descname">get_from_gtf_using_gene_name</tt><big>(</big><em>gtf</em>, <em>strand</em>, <em>chr</em>, <em>start</em>, <em>end</em><big>)</big><a class="reference internal" href="_modules/splice_graph.html#get_from_gtf_using_gene_name"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#splice_graph.get_from_gtf_using_gene_name" title="Permalink to this definition">¶</a></dt>
<dd><p>This function finds the first gene in the gtf that completely contains the target interval.
I should really think about checking for multiple genes instead of just returning.</p>
</dd></dl>

<dl class="function">
<dt id="splice_graph.get_sufficient_psi_exons">
<tt class="descclassname">splice_graph.</tt><tt class="descname">get_sufficient_psi_exons</tt><big>(</big><em>name</em>, <em>target</em>, <em>sGraph</em>, <em>genome</em>, <em>ID</em><big>)</big><a class="reference internal" href="_modules/splice_graph.html#get_sufficient_psi_exons"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#splice_graph.get_sufficient_psi_exons" title="Permalink to this definition">¶</a></dt>
<dd><p>Utilizes the ExonSeek class to find flanking exons that are
good enough to be called &#8220;constitutive&#8221;.</p>
</dd></dl>

<dl class="function">
<dt id="splice_graph.get_weakly_connected_tx">
<tt class="descclassname">splice_graph.</tt><tt class="descname">get_weakly_connected_tx</tt><big>(</big><em>gtf</em>, <em>strand</em>, <em>chr</em>, <em>start</em>, <em>end</em>, <em>plus_or_minus=1000000</em><big>)</big><a class="reference internal" href="_modules/splice_graph.html#get_weakly_connected_tx"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#splice_graph.get_weakly_connected_tx" title="Permalink to this definition">¶</a></dt>
<dd><p>This function is meant to handle tx annotations without gene ids.
Currently this is a function outside of the SpliceGraph class but it may be beneficial
to later include this as a method.</p>
</dd></dl>

<dl class="function">
<dt id="splice_graph.main">
<tt class="descclassname">splice_graph.</tt><tt class="descname">main</tt><big>(</big><em>options</em>, <em>args_output='tmp/debug.json'</em><big>)</big><a class="reference internal" href="_modules/splice_graph.html#main"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#splice_graph.main" title="Permalink to this definition">¶</a></dt>
<dd><p>The gtf main function is the function designed to be called from other scripts. It iterates through each target
exons and returns the necessary information for primer design.</p>
</dd></dl>

</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">splice_graph.py</a><ul>
<li><a class="reference internal" href="#splice-graph">Splice Graph</a></li>
<li><a class="reference internal" href="#transcripts-overlapping-target">Transcripts Overlapping Target</a></li>
<li><a class="reference internal" href="#flanking-exons">Flanking Exons</a></li>
<li><a class="reference internal" href="#documentation">Documentation</a></li>
</ul>
</li>
</ul>
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
  <li><a href="code.html">Code</a><ul>
      <li>Previous: <a href="primer.html" title="previous chapter">primer.py</a></li>
      <li>Next: <a href="algorithms.html" title="next chapter">algorithms.py</a></li>
  </ul></li>
  </ul></li>
</ul>
  <h3>This Page</h3>
  <ul class="this-page-menu">
    <li><a href="_sources/splice_graph.txt"
           rel="nofollow">Show Source</a></li>
  </ul>
<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy; Copyright 2012, CollinTokheim.
    </div>
    <a href="https://github.com/ctokheim/PrimerSeq" class="github">
        <img style="position: absolute; top: 0; right: 0; border: 0;" src="http://s3.amazonaws.com/github/ribbons/forkme_right_darkblue_121621.png" alt="Fork me on GitHub" />
    </a>
<!--
    <script type="text/javascript">

      var _gaq2 = _gaq2 || [];
      _gaq2.push(['_setAccount', 'UA-8742933-10']);
      _gaq2.push(['_setDomainName', 'none']);
      _gaq2.push(['_setAllowLinker', true]);
      _gaq2.push(['_trackPageview']);

      (function() {
        var ga2 = document.createElement('script'); ga.type = 'text/javascript'; ga2.async = true;
        ga2.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
        var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga2, s);
      })();

    </script>

    <script type="text/javascript">
      (function() {
        var t   = document.createElement('script');
        t.type  = 'text/javascript';
        t.async = true;
        t.id    = 'gauges-tracker';
        t.setAttribute('data-site-id',
                       '4ddc1cfaf5a1f50fcc000001');
        t.src = '//secure.gaug.es/track.js';
        var s = document.getElementsByTagName('script')[0];
        s.parentNode.insertBefore(t, s);
      })();
    </script>
-->
  </body>
</html>